apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/HeaderGetter.java

```diff
@@ -31,6 +31,23 @@
     @Nullable
     T getFirstHeader(String headerName, C carrier);
 
-    @Nullable
-    Iterable<T> getHeaders(String headerName, C carrier);
+    /**
+     * Calls the consumer for each header value with the given key
+     * until all entries have been processed or the action throws an exception.
+     * <p>
+     * The third parameter lets callers pass in a stateful object to be modified with header values,
+     * so the {@link HeaderConsumer} implementation itself can be stateless and potentially reusable.
+     * </p>
+     *
+     * @param headerName the name of the header
+     * @param carrier    the object containing the headers
+     * @param state      the object to be passed as the second parameter to each invocation on the specified consumer
+     * @param consumer   the action to be performed for each header value
+     * @param <S>        the type of the state object
+     */
+    <S> void forEach(String headerName, C carrier, S state, HeaderConsumer<T, S> consumer);
+
+    interface HeaderConsumer<T, S> {
+        void accept(T headerValue, S state);
```

Comment:
The consumer states I currently see are:
1. before header iteration
2. during header iteration
3. after header iteration

Since the caller (i.e. agent) is aware of these states and can easily make the consumer aware of them, I currently don't see a user for the state. What did you have in mind? Can you provide an example?

Reply:
OK, implementing now and realized what you had in mind 👍